File: C++/4.建造者模式/BuilderPatternDemo.cpp
/**
 * @file BuilderPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-10-23
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "MealBuilder.h"

int main(int argc, char const *argv[]) {
    MealBuilder meal_builder;

    std::shared_ptr<Meal> veg_meal = meal_builder.PrepareVegMeal();
    std::cout << "Veg Meal" << std::endl;
    veg_meal->ShowItems();
    std::cout << "Total Cost " << veg_meal->GetCost() << std::endl;

    std::shared_ptr<Meal> non_veg_meal = meal_builder.PrepareNonVegMeal();
    std::cout << "Veg Meal" << std::endl;
    non_veg_meal->ShowItems();
    std::cout << "Total Cost" << non_veg_meal->GetCost() << std::endl;

    return 0;
}

File: Objective-C/4.BuilderPattern/BuilderPatternDemo.m
//
//  4.BuilderPatternDemo.m
//  4.BuilderPattern
//
//  Created by Layne on 2020/10/18.
//

#import "MealBuilder.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        MealBuilder* mealBuilder = [[MealBuilder alloc] init];

        Meal* vegMeal = [mealBuilder prepareVegMeal];
        NSLog(@"Veg Meal");
        [vegMeal showItems];
        NSLog(@"Total Cost: %f", [vegMeal getCost]);

        Meal* nonVegMeal = [mealBuilder prepareNonVegMeal];
        NSLog(@"\n\nNon-Veg Meal");
        [nonVegMeal showItems];
        NSLog(@"Total Cost: %f", [nonVegMeal getCost]);
    }
    return 0;
}
